fix(mme): Handling Maximum number of PDU sessions reached( Cause #65)

ganeshg87 opened the pull request:
Test Cases:
    - verfied pdu session limit handling using ueransim
    - unit testcases added to validate functions

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

<!-- Enumerate changes you made and why you made them -->

## Test Plan

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

ganeshg87 committed:
fix(mme): Handling Maximum number of PDU sessions reached( Cause #65)

Test Cases:
    - verfied pdu session limit handling using ueransim
    - unit testcases added to validate functions

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

lionelgo commented on lte/gateway/c/core/oai/tasks/amf/amf_smf_send.cpp:
```diff
@@ -15,27 +15,34 @@
 #endif
 #include "include/amf_session_manager_pco.h"
 #include "amf_recv.h"
+#include "./amf_sap.h"
```
file path here ?

ganeshg87 replied:
yes, headers and sources are on the same path.
but I have defined for one of warning, it looks odd, removed

ganeshg87 commented:
verified UT and UERANSIM.
please find logs and pcap attached 
[mme_max_pdu.log](https://github.com/magma/magma/files/7156123/mme_max_pdu.log)
![image](https://user-images.githubusercontent.com/83060027/133126246-1f5d3ea7-48b2-482b-a879-78deb7e66902.png)
![image](https://user-images.githubusercontent.com/83060027/133126285-d9364b9d-e2ca-4dbe-977e-7f765a746234.png)

ulaskozat commented on lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GPayloadContainer.h:
```diff
@@ -12,13 +12,14 @@
 #include <cstdint>
+#include <string.h>
 #include <SmfMessage.h>
+#define PAYLOAD_CONTAINER_CONTENTS_MAX_LEN 8192
```
are use using namespace std in this file? if so please remove `using namespace std` and explicitly use `std::`

ulaskozat replied:
if you are not using it at all, simply remove it.

ganeshg87 replied:
the namespace is not using , removed it and updated review

ganeshg87 replied:
@ulaskozat,  I have resolved the above review comment for complete nas5g module. but too many changes
                   can I include it in the same review?

ulaskozat reviewed (changes requested):
please fix github action comments

ulaskozat commented on lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GPayloadContainer.cpp:
```diff
@@ -47,19 +47,25 @@ int PayloadContainerMsg::EncodePayloadContainerMsg(
 
-  return (encoded);
+  return (encoded + tmp);
 };
```
remove `;` after `}` as indicated by github actions.

ganeshg87 replied:
addressed comment.

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -517,6 +522,101 @@ TEST(test_amf_nas5g_pkt_process, test_amf_service_reject_message) {
+
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(buffer->data, &msg, len, nullptr);
+  EXPECT_TRUE(bytes > 0);
```
you should encode the message back and verify that it matches to the original message.

ganeshg87 replied:
addressed review comment

lionelgo commented on lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h:
```diff
@@ -29,6 +29,22 @@ class SmfMsgHeader {
+    message_type                    = s.message_type;
+  }
+  bool isEqual(const SmfMsgHeader& s) {
+    if ((extended_protocol_discriminator ==
```
Would suggest only bool isEqual(const SmfMsgHeader& s) const {

ganeshg87 replied:
but isEqual api calling directly with instance variables of Each class. the "const member function" caller should be a constant variable. I am getting the below errors.
![image](https://user-images.githubusercontent.com/83060027/134062526-60b77c57-f9c4-43bd-a119-a90b3d249c13.png)

lionelgo commented on lte/gateway/c/core/oai/tasks/nas5g/include/SmfMessage.h:
```diff
@@ -57,5 +73,28 @@ class SmfMsg {
+        break;
+    }
+  }
+  bool isEqual(const SmfMsg& s) {
```
Would suggest only bool isEqual(const SmfMsg& s) const {

ganeshg87 replied:
comment fix not resolved

lionelgo commented on lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GMessageType.h:
```diff
@@ -26,5 +26,13 @@ class MessageTypeMsg {
+    }
+    return false;
+  }
 };
```
I suggest  return msg_type == m.msg_type;

lionelgo replied:
and const method for all isEqual() methods

ganeshg87 replied:
addressed the first comment.
2nd comment dropped

ganeshg87 committed:
addressed review comments-R1

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ganeshg87 commented:
@ulaskozat, @lionelgo,

     addressed review comments.Please provide your approval.

    -  max PDU limit verified with UERANSIM.
    -  updated AMF UT as Ulas suggested and verified.
[mme_max_pdu_21_sept.log](https://github.com/magma/magma/files/7198390/mme_max_pdu_21_sept.log)
![image](https://user-images.githubusercontent.com/83060027/134066775-08d962db-2c22-4194-8d80-24a14905c39e.png)
![image](https://user-images.githubusercontent.com/83060027/134067308-c7e3cfea-ad4b-4a59-a3ee-919829691bae.png)

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -24,6 +24,11 @@ extern "C" {
+#define QUADLET 4
+#define AMF_GET_BYTE_ALIGNED_LENGTH(LENGTH)                                    \
+  LENGTH += QUADLET - (LENGTH % QUADLET)
+#endif
```
why do you have this?

ganeshg87 replied:
addressed review comment

ulaskozat commented on lte/gateway/c/core/oai/test/amf/test_amf.cpp:
```diff
@@ -517,6 +522,103 @@ TEST(test_amf_nas5g_pkt_process, test_amf_service_reject_message) {
+
+  buffer = bfromcstralloc(len, "\0");
+  bytes  = nas5g_message_encode(buffer->data, &msg, len, nullptr);
+  EXPECT_TRUE(bytes > 0);
```
legit github actions comment

ganeshg87 replied:
addressed review comment

ganeshg87 committed:
addressed review comments-R2

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

ulaskozat reviewed (approved):
LGTM
